File: python/trees/lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        ancestors = []

        def dfs(node):
            hasp, hasq = False, False
            if not node:
                return hasp, hasq

            hasp, hasq = dfs(node.left)
            if hasp:
                _, hasq = dfs(node.right)
            elif hasq:
                hasp, _ = dfs(node.right)
            else:
                hasp, hasq = dfs(node.right)

            if node == p:
                hasp = True
            if node == q:
                hasq = True

            if hasp and hasq:
                ancestors.append(node)

            return hasp, hasq

        dfs(root)
        return ancestors[0]

File: c++/trees/lowest-common-ancestor-of-a-binary-search-tree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return std::get<0>(find(root, p, q));
    }

private:
    std::tuple<TreeNode*, bool, bool> find(TreeNode* node, TreeNode* p, TreeNode* q) {
        if (!node) {
            return std::make_tuple(nullptr, false, false);
        }

        std::tuple<TreeNode*, bool, bool> left = find(node->left, p, q);
        std::tuple<TreeNode*, bool, bool> right = find(node->right, p, q);

        if (std::get<0>(left)) {
            return std::make_tuple(std::get<0>(left), true, true);
        }

        if (std::get<0>(right)) {
            return std::make_tuple(std::get<0>(right), true, true);
        }

        if (node == p) {
            if (std::get<2>(left) || std::get<2>(right)) {
                return std::make_tuple(node, true, true);
            }
        }

        if (node == q) {
            if (std::get<1>(left) || std::get<1>(right)) {
                return std::make_tuple(node, true, true);
            }
        }

        if ((std::get<2>(left) || std::get<2>(right)) && (std::get<1>(left) || std::get<1>(right))) {
            return std::make_tuple(node, true, true);
        }

        return std::make_tuple(nullptr, std::get<1>(left) || std::get<1>(right) || node == p, std::get<2>(left) || std::get<2>(right) || node == q);
    }
};
